cluster_experiments/perturbator.py

```diff
@@ -350,6 +351,266 @@ def _sample_beta_effect(
+        return x * (self._range_max - self._range_min) + self._range_min
+
+    def _inv_transform_to_range(self, x: Union[float, np.ndarray]):
+        return (x - self._range_min) / (self._range_max - self._range_min)
+
+    @classmethod
+    def from_config(cls, config):
+        """Creates a Perturbator object from a PowerConfig object"""
+        return cls(
+            average_effect=config.average_effect,
+            target_col=config.target_col,
+            treatment_col=config.treatment_col,
+            treatment=config.treatment,
+            scale=config.scale,
+            range_min=config.range_min,
+            range_max=config.range_max,
+        )
+
+
+class ClusteredBetaRelativePerturbator(BetaRelativePositivePerturbator):
+    """
+    A stochastic Perturbator for continuous targets that applies a sampled
+    effect from the Beta distribution. It applies the effect multiplicatively
+    and based on given clusters.
+    For each cluster, the average cluster effect is sampled from a beta
+    distribution with support in (0, 1). Within each cluster, the individual
+    effects are sampled from a beta distribution with mean equal to the cluster
+    average effect and support in (range_min, range_max).
+
+    The number of samples with 0 as target remains unchanged.
+
+    Arguments:
+        average_effect (Optional[float], optional): the average effect of the treatment. Defaults to None.
+        target_col (str, optional): name of the target_col to use as the outcome. Defaults to "target".
+        treatment_col (str, optional): the name of the column that contains the treatment. Defaults to "treatment".
+        treatment (str, optional): name of the treatment to use as the treated group. Defaults to "B".
+        scale (Optional[float], optional): the scale of the effect distribution. Defaults to None.
+        range_min (float, optional): the minimum value of the target range. Defaults to -0.8.
+        range_max (float, optional): the maximum value of the target range. Defaults to 5.
+        cluster_cols (Optional[List[str]], optional): the columns to use for clustering. Defaults to None.
+    """
+
+    def __init__(
+        self,
+        cluster_cols: List[str],
+        average_effect: Optional[float] = None,
+        target_col: str = "target",
+        treatment_col: str = "treatment",
+        treatment: str = "B",
+        scale: Optional[float] = None,
+        range_min: float = -0.8,
+        range_max: float = 5,
+    ):
+        super().__init__(average_effect, target_col, treatment_col, treatment, scale)
+        self._range_min = range_min
+        self._range_max = range_max
+        self._cluster_cols = cluster_cols
+        self.cluster_col = self._get_cluster_col_name(cluster_cols)
+
+    def get_cluster_perturbator(
+        self, average_effect: Optional[float] = None
+    ) -> Perturbator:
+        average_effect = self.get_average_effect(average_effect)
```

Comment:
do you need to do all of this for each perturbator? or only once at the perturbate method?

Reply:
Hmm yeah you're right that much of this only has to be done once. I now separated out a `get_cluster_perturbator_fixed_params` method which is only run once, and then the `get_cluster_perturbator` is run for each cluster. It may make the code a bit more wordy, but if ever we want to create another clustered perturbator, we only have to adjust few methods and it should be clear which ones to adjust. I hope this goes in the direction of what you meant here:)